src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -182,6 +189,94 @@ int MPL_gavl_tree_create(void (*free_fn) (void *), MPL_gavl_tree_t * gavl_tree)
     goto fn_exit;
 }
 
+static void gavl_tree_insert_search(gavl_tree_s * gavl_tree_iptr, gavl_tree_node_s * node_ptr,
+                                    gavl_tree_node_s * cur_node, int *direction,
+                                    gavl_tree_node_s ** cur_node_ret,
+                                    gavl_tree_stack_s * node_stack)
+{
+    int cmp_ret;
+
+    do {
+        cmp_ret = gavl_subset_cmp_func((uintptr_t) node_ptr->addr, node_ptr->len, cur_node);
+        if (cmp_ret == SEARCH_LEFT) {
+            if (cur_node->left != NULL) {
+                STACK_PUSH(node_stack, cur_node);
+                cur_node = cur_node->left;
+                continue;
+            } else {
+                *direction = SEARCH_LEFT;
+                break;
+            }
+        } else if (cmp_ret == SEARCH_RIGHT) {
+            if (cur_node->right != NULL) {
+                STACK_PUSH(node_stack, cur_node);
+                cur_node = cur_node->right;
+                continue;
+            } else {
+                *direction = SEARCH_RIGHT;
+                break;
+            }
+        } else {
+            /* we cannot insert the duplicate buffer */
+            gavl_tree_iptr->gavl_free_fn((void *) node_ptr->val);
+            MPL_free(node_ptr);
+            *direction = NO_BUFFER_MATCH;
+            break;
+        }
+    } while (1);
+
+    *cur_node_ret = cur_node;
+    return;
+}
+
+static void gavl_tree_insert_node(int direction, gavl_tree_node_s * node_ptr,
+                                  gavl_tree_node_s * cur_node)
+{
+    if (direction == SEARCH_LEFT)
+        cur_node->left = node_ptr;
+    else
+        cur_node->right = node_ptr;
+    node_ptr->parent = cur_node;
+    return;
+}
+
+static void gavl_tree_rebalance(gavl_tree_s * gavl_tree_iptr, gavl_tree_node_s * cur_node,
+                                gavl_tree_stack_s * node_stack)
```

Comment:
Again I don't understand why you need the stack or even the cur_node. If you need them as part of tree functions, then you should put them as part of the tree structure and have tree functions use them internally rather than passing them around. It is a burden to maintain them by caller.

Reply:
`cur_node` is needed because I need to know which node to start my rebalance process. 
`node_stack` records all traces when we call `gavl_tree_insert_search` and traverse the tree, the traces are used to find which node needs to be rebalanced. 
It is a design choice whether we should put `cur_node` in tree structure, I feel passing it through function call can make logic clearer.

Reply:
> It is a design choice whether we should put `cur_node` in tree structure, I feel passing it through function call can make logic clearer.

Think in semantics -- The code calling tree_search intuitively will need the `tree_ptr`, `addr + len`, and will expect return of a `node`. There is no context for "stack" or "cur_pointer". Those are internal scaffold and should be handled internally. Exposing them to parameter, you are weaving the internal logic with the higher layer semantics, resulting in more complex code. Now as we trying to maintain the higher level code, not only we need understand what we are doing and what we need, we also need understand the internal scaffold -- thus unnecessarily makes the code more difficult to understand rather than "clearer".